mypyc/ir/rtypes.py

```diff
@@ -820,17 +820,15 @@ def make_simplified_union(items: list[RType]) -> RType:
         items = flatten_nested_unions(items)
         assert items
 
-        # Remove duplicate items using set + list to preserve item order
-        seen = set()
-        new_items = []
+        # Use a dict for O(1) lookups that preserve order; values are ignored
+        seen: dict[RType, Any] = {}
         for item in items:
             if item not in seen:
-                new_items.append(item)
-                seen.add(item)
-        if len(new_items) > 1:
-            return RUnion(new_items)
+                seen[item] = None
```

Comment:
This could be written more simply as:

```py
seen = dict.fromkeys(items)
```

Reply:
Just pushed this, I swear you read my mind :)